zap/src/main/java/org/parosproxy/paros/core/scanner/VariantMultipartFormParameters.java

```diff
@@ -209,33 +212,78 @@ public void setMessage(HttpMessage msg) {
+    public void setParameters(HttpMessage msg, List<InputVector> inputVectors) {
+        this.setParameter(
+                msg,
+                inputVectors.stream()
+                        .map(inputVector -> inputVector.getPosition())
+                        .collect(Collectors.toList()),
+                inputVectors.stream()
+                        .map(inputVector -> inputVector.getValue())
+                        .collect(Collectors.toList()));
+    };
 
+    private String setParameter(
+            HttpMessage msg, List<Integer> nameValuePairPositions, List<String> values) {
         StringBuilder newBodyBuilder = new StringBuilder(msg.getRequestBody().toString());
-        int idx = originalPair.getPosition() - 1;
-        MultipartFormParameter mpPart = multiPartParams.get(idx);
-        if (LOGGER.isDebugEnabled()) {
-            LOGGER.debug(
-                    "i: "
-                            + idx
-                            + " pos: "
-                            + originalPair.getPosition()
-                            + " S: "
-                            + mpPart.getStart()
-                            + " E: "
-                            + mpPart.getEnd());
-        }
-        newBodyBuilder.replace(mpPart.getStart(), mpPart.getEnd(), value);
+        List<MultipartFormParameter> multiPartParamsClone =
+                new ArrayList<MultipartFormParameter>(this.multiPartParams);
+        for (int index = 0; index < nameValuePairPositions.size(); index++) {
+            int originalPosition = nameValuePairPositions.get(index);
+            String value = values.get(index);
+            int idx = originalPosition - 1;
 
+            MultipartFormParameter mpPart = multiPartParamsClone.get(idx);
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug(
+                        "i: "
+                                + idx
+                                + " pos: "
+                                + originalPosition
+                                + " S: "
+                                + mpPart.getStart()
+                                + " E: "
+                                + mpPart.getEnd());
+            }
+            newBodyBuilder.replace(mpPart.getStart(), mpPart.getEnd(), value);
+            int originalStart = mpPart.getStart();
+            int offset = value.length() - mpPart.getEnd() + mpPart.getStart();
+            mpPart =
+                    new MultipartFormParameter(
+                            mpPart.getName(),
+                            mpPart.getValue(),
+                            originalStart,
+                            mpPart.getEnd() + offset,
+                            mpPart.getPosition(),
+                            mpPart.getType());
+            multiPartParamsClone.set(idx, mpPart);
+            // As we have updated one value and this can impact other multipart params
+            // Hence this logic is to update the other multipart params
```

Comment:
It just needs to keep the offset and sum to existing positions. No need to copy the parameters again.

Reply:
in the for loop? actually, the setter changes are not there hence i cannot set the values inside the multipartformparameter hence need to create a new instance with newer values.

Reply:
can you please elaborate?

Reply:
Calculate the difference of the values injected to the ones already present then you can just offset (i.e. sum) the start/end of the following parameters when replacing.

Reply:
but to do that we don't have setters, we need to create again the object.
I think we should add the setters as that will make it easier to modify the values.

Reply:
@thc202 can you please help with this.

Reply:
are you suggesting to add a Map of position vs offset? in the start i will have offset as 0 then once i start modifying, i find the offset in the map and then while replacing use the offset.

cannot we create a setter with package private visibility?

Reply:
I tried map thing and it is making the logic complex. we can easily do this with setters. @thc202

Reply:
It just needs an int for the offset, I'll get to you later.

Reply:
This will need #6690.

Reply:
With the changes in place this can now be:
```diff
-        List<MultipartFormParameter> multiPartParamsClone =
-                new ArrayList<MultipartFormParameter>(this.multiPartParams);
+        int offset = 0;
         for (int index = 0; index < nameValuePairPositions.size(); index++) {
             int originalPosition = nameValuePairPositions.get(index);
             String value = values.get(index);
             int idx = originalPosition - 1;
 
-            MultipartFormParameter mpPart = multiPartParamsClone.get(idx);
+            MultipartFormParameter mpPart = multiPartParams.get(idx);
             if (LOGGER.isDebugEnabled()) {
                 LOGGER.debug(
                         "i: "
@@ -268,34 +269,8 @@ public class VariantMultipartFormParameters implements Variant {
                                 + " E: "
                                 + mpPart.getEnd());
             }
-            newBodyBuilder.replace(mpPart.getStart(), mpPart.getEnd(), value);
-            int originalStart = mpPart.getStart();
-            int offset = value.length() - mpPart.getEnd() + mpPart.getStart();
-            mpPart =
-                    new MultipartFormParameter(
-                            mpPart.getName(),
-                            mpPart.getValue(),
-                            originalStart,
-                            mpPart.getEnd() + offset,
-                            mpPart.getPosition(),
-                            mpPart.getType());
-            multiPartParamsClone.set(idx, mpPart);
-            // As we have updated one value and this can impact other multipart params
-            // Hence this logic is to update the other multipart params
-            for (idx = 0; idx < multiPartParamsClone.size(); idx++) {
-                mpPart = multiPartParamsClone.get(idx);
-                if (mpPart.getStart() > originalStart) {
-                    mpPart =
-                            new MultipartFormParameter(
-                                    mpPart.getName(),
-                                    mpPart.getValue(),
-                                    mpPart.getStart() + offset,
-                                    mpPart.getEnd() + offset,
-                                    mpPart.getPosition(),
-                                    mpPart.getType());
-                    multiPartParamsClone.set(idx, mpPart);
-                }
-            }
+            newBodyBuilder.replace(mpPart.getStart() + offset, mpPart.getEnd() + offset, value);
+            offset += value.length() - (mpPart.getEnd() - mpPart.getStart());
```

Reply:
Note that some tests will need to update the position of the `NameValuePair`s.

Reply:
This one was not addressed yet.

Reply:
Oh missed your comment. Will address

Reply:
I think this code is working because of sorting by position. Actually, I didn't do this earlier and used ArrayList to store the offset because of the unordering thing.